alternator/conditions.cc

```diff
@@ -143,9 +143,50 @@ static void verify_operand_count(const rjson::value* array, const size_check& ex
     }
 }
 
+struct rjson_engaged_ptr_comp {
+    bool operator()(const rjson::value* p1, const rjson::value* p2) const {
+        return rjson::single_value_comp()(*p1, *p2);
+    }
+};
+
+// It's not enough to compare underlying JSON objects when comparing sets,
+// as internally they're stored in an array, and the order of elements is
+// not important in set equality. See issue #5021
+static bool check_EQ_for_sets(const rjson::value& s1, const rjson::value& s2) {
+    const rjson::value* set1 = unwrap_set(s1).second;
+    const rjson::value* set2 = unwrap_set(s2).second;
+
+    if (!set1 || !set2) {
+        return !set1 && !set2;
+    }
+    if (set1->Size() != set2->Size()) {
+        return false;
+    }
+    std::set<const rjson::value*, rjson_engaged_ptr_comp> set1_raw;
+    for (auto it = set1->Begin(); it != set1->End(); ++it) {
+        set1_raw.insert(&*it);
```

Comment:
As I commented earlier, if this insert() sees an item with the wrong type, it can cause an "internal server error" instead of a normal invlaidation. This is not great (although I agree nobody will ever reach this case anyway in a real application.